include/swift/AST/Identifier.h

```diff
@@ -203,7 +205,144 @@ namespace llvm {
+/// version of both SwiftLookupTable AND ModuleFormat must thus be incremented.
+class DeclBaseName {
+public:
+  enum class Kind : uint8_t {
+    Normal = 0,
+    Subscript
+  };
   
+private:
+  /// In a special DeclName represenenting a subscript, this opaque pointer
+  /// is used as the data of the base name identifier.
+  /// This is an implementation detail that should never leak outside of
+  /// DeclName.
+  static void *SubscriptIdentifierData;
+
+  Identifier Ident;
+
+public:
+  DeclBaseName() : DeclBaseName(Identifier()) {}
+
+  DeclBaseName(Identifier I) : Ident(I) {}
+
+  static DeclBaseName createSubscript() {
+    return DeclBaseName(Identifier((const char *)SubscriptIdentifierData));
+  }
+
+  Kind getKind() const {
+    if (Ident.get() == SubscriptIdentifierData) {
+      return Kind::Subscript;
+    } else {
+      return Kind::Normal;
+    }
+  }
+
+  bool isSpecial() const { return getKind() != Kind::Normal; }
+
+  /// Return the identifier backing the name. Assumes that the name is not
+  /// special.
+  Identifier getIdentifier() const {
+    assert(!isSpecial() && "Cannot retrieve identifier from special names");
+    return Ident;
+  }
+
+  bool empty() const { return !isSpecial() && getIdentifier().empty(); }
+
+  bool isOperator() const {
+    return !isSpecial() && getIdentifier().isOperator();
+  }
+
+  bool isEditorPlaceholder() const {
+    return !isSpecial() && getIdentifier().isEditorPlaceholder();
+  }
+
+  /// A representation of the name to be displayed to users. May be ambiguous
+  /// between identifiers and special names
+  StringRef userFacingStr() const {
+    switch (getKind()) {
+    case Kind::Normal:
+      return getIdentifier().str();
+    case Kind::Subscript:
+      return "subscript";
+    }
+  }
+
+  int compare(DeclBaseName other) const {
+    return userFacingStr().compare(other.userFacingStr());
+  }
+
+  bool operator==(StringRef Str) const {
+    return !isSpecial() && getIdentifier().str() == Str;
+  }
+  bool operator!=(StringRef Str) const { return !(*this == Str); }
+
+  bool operator==(DeclBaseName RHS) const { return Ident == RHS.Ident; }
+  bool operator!=(DeclBaseName RHS) const { return !(*this == RHS); }
+
+  bool operator<(DeclBaseName RHS) const {
+    return Ident.get() < RHS.Ident.get();
+  }
+
+  const void *getAsOpaquePointer() const { return Ident.get(); }
+
+  static DeclBaseName getFromOpaquePointer(void *P) {
+    return Identifier::getFromOpaquePointer(P);
+  }
+
+  static DeclBaseName getEmptyKey() { return Identifier::getEmptyKey(); }
+  static DeclBaseName getTombstoneKey() {
+    return Identifier::getTombstoneKey();
+  }
+};
+
+} // end namespace swift
+
+namespace llvm {
+
+raw_ostream &operator<<(raw_ostream &OS, swift::DeclBaseName D);
+
+// Identifiers hash just like pointers.
+template<> struct DenseMapInfo<swift::DeclBaseName> {
+  static swift::DeclBaseName getEmptyKey() {
+    return swift::DeclBaseName::getEmptyKey();
```

Comment:
Nitpick: Why the extra level of indirection? It's not like `getEmptyKey` is going to be used anywhere else.

(I see that Identifier did this too, but it's weird there too.)